AKS failure might be #16938 

>  ❌ 6/11 tests failed (1/84 actions), 0 tests skipped, 0 scenarios skipped:
> Test [to-entities-world]:
>   ❌ to-entities-world/pod-to-world/https-to-one-one-one-one-index-1: cilium-test/client2-6dd75b74c6-8pjfx (10.240.0.90) -> one-one-one-one-https-index (one.one.one.one:443)
> Test [to-cidr-1111]:
> Test [echo-ingress-l7]:
> Test [client-egress-l7]:
> Test [dns-only]:
> Test [to-fqdns]:
> Connectivity test failed: 6 tests failed
> Error: Process completed with exit code 1.

--- a/pkg/datapath/linux/config/config.go
+++ b/pkg/datapath/linux/config/config.go
@@ -231,6 +231,10 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 		cDefinesMap["ENABLE_EGRESS_GATEWAY"] = "1"
 	}
 
+	if option.Config.EnableEndpointRoutes {
+		cDefinesMap["ENABLE_ENDPOINT_ROUTES"] = "1"
+	}
+
 	if option.Config.EnableHostReachableServices {
 		if option.Config.EnableHostServicesTCP {
 			cDefinesMap["ENABLE_HOST_SERVICES_TCP"] = "1"
@@ -816,10 +820,6 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 
 	if e.RequireRouting() {
 		fmt.Fprintf(fw, "#define ENABLE_ROUTING 1\n")
-	}
-
-	if e.RequireEndpointRoute() {
-		fmt.Fprintf(fw, "#define ENABLE_ENDPOINT_ROUTES 1\n")
 	}
 
 	if e.DisableSIPVerification() {

--- a/pkg/datapath/linux/ipsec/ipsec_linux.go
+++ b/pkg/datapath/linux/ipsec/ipsec_linux.go
@@ -20,6 +20,7 @@ import (
 	"github.com/cilium/cilium/pkg/datapath/linux/linux_defaults"
 	"github.com/cilium/cilium/pkg/datapath/linux/route"
 	"github.com/cilium/cilium/pkg/maps/encrypt"
+	"github.com/cilium/cilium/pkg/option"
 
 	"github.com/sirupsen/logrus"
 	"github.com/vishvananda/netlink"
@@ -160,12 +161,17 @@ func _ipSecReplacePolicyInFwd(src, dst, tmplSrc, tmplDst *net.IPNet, tunnel bool
 		policy.Mark = &netlink.XfrmMark{
 			Mask: linux_defaults.IPsecMarkMaskIn,
 		}
-		if tunnel {
-			// Required as this policy with the following mark does not have a
-			// corresponding XFRM state matching it. The XFRM state for Dir=In
-			// has mark for decryption only. If we don't mark this optional,
-			// then we'll get a packet drop with the reason as
-			// XfrmInTmplMismatch.
+		if tunnel || option.Config.EnableEndpointRoutes {
+			// Required for tunneling mode as this policy with the following
+			// mark does not have a corresponding XFRM state matching it. The
+			// XFRM state for Dir=In has mark for decryption only. If we don't
+			// mark this optional, then we'll get a packet drop with the
+			// reason as XfrmInTmplMismatch.
+			// Required for endpoint routes because packets may have either the
+			// decrypt or the proxy mark when attempting to match them against
+			// XFRM policies, depending on whether the connection goes through
+			// the proxy. If not marked optional, packets are dropped with
+			// XfrmInNoPols.
 			optional = 1
 			policy.Mark.Value = linux_defaults.RouteMarkToProxy
 		} else {

--- a/pkg/datapath/linux/node.go
+++ b/pkg/datapath/linux/node.go
@@ -967,7 +967,7 @@ func (n *linuxNodeHandler) enableIPsec(newNode *nodeTypes.Node) {
 
 				/* Insert wildcard policy rules for traffic skipping back through host */
 				ipsecIPv4Wildcard := &net.IPNet{IP: net.ParseIP(wildcardIPv4), Mask: net.IPv4Mask(0, 0, 0, 0)}
-				if err = ipsec.IpSecReplacePolicyFwd(ipsecIPv4Wildcard, ipsecRemote, ipsecLocal, ipsecRemote); err != nil {
+				if err = ipsec.IpSecReplacePolicyFwd(ipsecIPv4Wildcard, ipsecRemote, ipsecIPv4Wildcard, ipsecRemote); err != nil {
 					log.WithError(err).Warning("egress unable to replace policy fwd:")
 				}
 			}

